From: Sam Ravnborg <sam@ravnborg.org>
Date: Thu, 6 Sep 2018 22:07:48 +0200
Subject: [PATCH] imx6-cpu: add device tree files

Squash all previous commits to devicetree to make patch
maintenance simpler.

Signed-off-by: Sam Ravnborg <sam@ravnborg.org>
---
 arch/arm/boot/dts/Makefile                   |   4 +
 arch/arm/boot/dts/imx6dl-skov-revc-lt2.dts   |  23 +
 arch/arm/boot/dts/imx6dl-skov-revc-lt6.dts   |  27 ++
 arch/arm/boot/dts/imx6q-skov-revc-lt2.dts    |  36 ++
 arch/arm/boot/dts/imx6q-skov-revc-lt6.dts    |  40 ++
 arch/arm/boot/dts/imx6qdl-skov-cpu-revc.dtsi |  80 ++++
 arch/arm/boot/dts/imx6qdl-skov-cpu.dtsi      | 644 +++++++++++++++++++++++++++
 7 files changed, 854 insertions(+)
 create mode 100644 arch/arm/boot/dts/imx6dl-skov-revc-lt2.dts
 create mode 100644 arch/arm/boot/dts/imx6dl-skov-revc-lt6.dts
 create mode 100644 arch/arm/boot/dts/imx6q-skov-revc-lt2.dts
 create mode 100644 arch/arm/boot/dts/imx6q-skov-revc-lt6.dts
 create mode 100644 arch/arm/boot/dts/imx6qdl-skov-cpu-revc.dtsi
 create mode 100644 arch/arm/boot/dts/imx6qdl-skov-cpu.dtsi

diff --git a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
index b5bd3de87c33..dc0f9e3fec5f 100644
--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -424,6 +424,8 @@ dtb-$(CONFIG_SOC_IMX6Q) += \
 	imx6dl-sabrelite.dtb \
 	imx6dl-sabresd.dtb \
 	imx6dl-savageboard.dtb \
+	imx6dl-skov-revc-lt2.dtb \
+	imx6dl-skov-revc-lt6.dtb \
 	imx6dl-ts4900.dtb \
 	imx6dl-ts7970.dtb \
 	imx6dl-tx6dl-comtft.dtb \
@@ -500,6 +502,8 @@ dtb-$(CONFIG_SOC_IMX6Q) += \
 	imx6q-sabresd.dtb \
 	imx6q-savageboard.dtb \
 	imx6q-sbc6x.dtb \
+	imx6q-skov-revc-lt2.dtb \
+	imx6q-skov-revc-lt6.dtb \
 	imx6q-tbs2910.dtb \
 	imx6q-ts4900.dtb \
 	imx6q-ts7970.dtb \
diff --git a/arch/arm/boot/dts/imx6dl-skov-revc-lt2.dts b/arch/arm/boot/dts/imx6dl-skov-revc-lt2.dts
new file mode 100644
index 000000000000..f987973bc7f6
--- /dev/null
+++ b/arch/arm/boot/dts/imx6dl-skov-revc-lt2.dts
@@ -0,0 +1,23 @@
+/*
+ * Copyright 2014 Rose Technology
+ * Copyright 2012 Freescale Semiconductor, Inc.
+ * Copyright 2011 Linaro Ltd.
+ *
+ * The code contained herein is licensed under the GNU General Public
+ * License. You may obtain a copy of the GNU General Public License
+ * Version 2 or later at the following locations:
+ *
+ * http://www.opensource.org/licenses/gpl-license.html
+ * http://www.gnu.org/copyleft/gpl.html
+ */
+
+/dts-v1/;
+#include <dt-bindings/interrupt-controller/irq.h>
+#include "imx6dl.dtsi"
+#include "imx6qdl-skov-cpu.dtsi"
+#include "imx6qdl-skov-cpu-revc.dtsi"
+
+/ {
+	model = "SKOV IMX6 CPU SoloCore";
+	compatible = "skov,imx6dl-skov-revc-lt2", "skov,imx6", "fsl,imx6dl";
+};
diff --git a/arch/arm/boot/dts/imx6dl-skov-revc-lt6.dts b/arch/arm/boot/dts/imx6dl-skov-revc-lt6.dts
new file mode 100644
index 000000000000..cc278ac3f435
--- /dev/null
+++ b/arch/arm/boot/dts/imx6dl-skov-revc-lt6.dts
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2014 Rose Technology
+ * Copyright 2012 Freescale Semiconductor, Inc.
+ * Copyright 2011 Linaro Ltd.
+ *
+ * The code contained herein is licensed under the GNU General Public
+ * License. You may obtain a copy of the GNU General Public License
+ * Version 2 or later at the following locations:
+ *
+ * http://www.opensource.org/licenses/gpl-license.html
+ * http://www.gnu.org/copyleft/gpl.html
+ */
+
+/dts-v1/;
+#include <dt-bindings/interrupt-controller/irq.h>
+#include "imx6dl.dtsi"
+#include "imx6qdl-skov-cpu.dtsi"
+#include "imx6qdl-skov-cpu-revc.dtsi"
+
+/ {
+	model = "SKOV IMX6 CPU SoloCore";
+	compatible = "skov,imx6dl-skov-revc-lt6", "skov,imx6", "fsl,imx6dl";
+};
+
+&panel {
+	compatible = "logictechnologies,lttd800480070-l6wh-rt";
+};
diff --git a/arch/arm/boot/dts/imx6q-skov-revc-lt2.dts b/arch/arm/boot/dts/imx6q-skov-revc-lt2.dts
new file mode 100644
index 000000000000..7132c1fd6094
--- /dev/null
+++ b/arch/arm/boot/dts/imx6q-skov-revc-lt2.dts
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2014 Rose Technology
+ * Copyright 2012 Freescale Semiconductor, Inc.
+ * Copyright 2011 Linaro Ltd.
+ *
+ * The code contained herein is licensed under the GNU General Public
+ * License. You may obtain a copy of the GNU General Public License
+ * Version 2 or later at the following locations:
+ *
+ * http://www.opensource.org/licenses/gpl-license.html
+ * http://www.gnu.org/copyleft/gpl.html
+ */
+
+/dts-v1/;
+#include <dt-bindings/interrupt-controller/irq.h>
+#include "imx6q.dtsi"
+#include "imx6qdl-skov-cpu.dtsi"
+#include "imx6qdl-skov-cpu-revc.dtsi"
+
+/ {
+	model = "SKOV IMX6 CPU QuadCore";
+	compatible = "skov,imx6q-skov-revc-lt2", "skov,imx6", "fsl,imx6q";
+};
+
+&sata {
+	status = "disabled";
+};
+
+&i2c2 {
+	status = "okay";
+};
+
+&hdmi {
+	status = "okay";
+	ddc-i2c-bus = <&i2c2>;
+};
diff --git a/arch/arm/boot/dts/imx6q-skov-revc-lt6.dts b/arch/arm/boot/dts/imx6q-skov-revc-lt6.dts
new file mode 100644
index 000000000000..07dd8bea7f24
--- /dev/null
+++ b/arch/arm/boot/dts/imx6q-skov-revc-lt6.dts
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2014 Rose Technology
+ * Copyright 2012 Freescale Semiconductor, Inc.
+ * Copyright 2011 Linaro Ltd.
+ *
+ * The code contained herein is licensed under the GNU General Public
+ * License. You may obtain a copy of the GNU General Public License
+ * Version 2 or later at the following locations:
+ *
+ * http://www.opensource.org/licenses/gpl-license.html
+ * http://www.gnu.org/copyleft/gpl.html
+ */
+
+/dts-v1/;
+#include <dt-bindings/interrupt-controller/irq.h>
+#include "imx6q.dtsi"
+#include "imx6qdl-skov-cpu.dtsi"
+#include "imx6qdl-skov-cpu-revc.dtsi"
+
+/ {
+	model = "SKOV IMX6 CPU QuadCore";
+	compatible = "skov,imx6q-skov-revc-lt6", "skov,imx6", "fsl,imx6q";
+};
+
+&sata {
+	status = "disabled";
+};
+
+&i2c2 {
+	status = "okay";
+};
+
+&hdmi {
+	status = "okay";
+	ddc-i2c-bus = <&i2c2>;
+};
+
+&panel {
+	compatible = "logictechnologies,lttd800480070-l6wh-rt";
+};
diff --git a/arch/arm/boot/dts/imx6qdl-skov-cpu-revc.dtsi b/arch/arm/boot/dts/imx6qdl-skov-cpu-revc.dtsi
new file mode 100644
index 000000000000..97284a974c1c
--- /dev/null
+++ b/arch/arm/boot/dts/imx6qdl-skov-cpu-revc.dtsi
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2014 Rose Technology.
+ * Copyright 2012 Freescale Semiconductor, Inc.
+ * Copyright 2011 Linaro Ltd.
+ *
+ * The code contained herein is licensed under the GNU General Public
+ * License. You may obtain a copy of the GNU General Public License
+ * Version 2 or later at the following locations:
+ *
+ * http://www.opensource.org/licenses/gpl-license.html
+ * http://www.gnu.org/copyleft/gpl.html
+ */
+
+&pinctrl_board {
+	pinctrl_flexcan2_1: flexcan2grp-1 {
+		fsl,pins = <
+			MX6QDL_PAD_KEY_COL4__FLEXCAN2_TX 	0x80000000
+			MX6QDL_PAD_KEY_ROW4__FLEXCAN2_RX 	0x80000000
+		>;
+	};
+
+	pinctrl_ads7846: ads7846grp {
+		fsl,pins = <
+			/* external pull up */
+			MX6QDL_PAD_EIM_D19__GPIO3_IO19		0x10040		/* TSC2046 Pen down irq */
+		>;
+	};
+
+	pinctrl_backlight: backlight_grp {
+		fsl,pins = <
+			MX6QDL_PAD_RGMII_TD3__GPIO6_IO23	0x40000058
+		>;
+	};
+
+	pinctrl_vcc_mmc: vcc_mmc_grp {
+		fsl,pins = <
+			MX6QDL_PAD_SD3_RST__GPIO7_IO08		0x40000058	/* MMC Power Supply Switch */
+		>;
+	};
+};
+
+&backlight_lcd {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_backlight>;
+	enable-gpios = <&gpio6 23 GPIO_ACTIVE_LOW>;
+};
+
+&gpio6 {
+	backlight {
+		gpios = <23 0>; /* initially switch off the panel backlight */
+		output-high;
+		gpio-initval;
+	};
+};
+
+&ads7846 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_ads7846>;
+	pendown-gpio = <&gpio3 19 0>;
+	interrupt-parent = <&gpio3>;
+	interrupts = <19 IRQ_TYPE_LEVEL_LOW>;
+};
+
+&can2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_flexcan2_1>;
+	xceiver-supply = <&reg_3p3v>;
+	status = "okay";
+};
+
+&vcc_mmc {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_vcc_mmc>;
+	gpio = <&gpio7 8 GPIO_ACTIVE_HIGH>;
+	enable-active-high;
+	startup-delay-us = <100>;
+	/* don't switch off the SD card automatically */
+	regulator-always-on;
+	regulator-boot-on;
+};
diff --git a/arch/arm/boot/dts/imx6qdl-skov-cpu.dtsi b/arch/arm/boot/dts/imx6qdl-skov-cpu.dtsi
new file mode 100644
index 000000000000..8dbbddbd0aac
--- /dev/null
+++ b/arch/arm/boot/dts/imx6qdl-skov-cpu.dtsi
@@ -0,0 +1,644 @@
+/*
+ * Copyright 2014 Rose Technology.
+ * Copyright 2012 Freescale Semiconductor, Inc.
+ * Copyright 2011 Linaro Ltd.
+ *
+ * The code contained herein is licensed under the GNU General Public
+ * License. You may obtain a copy of the GNU General Public License
+ * Version 2 or later at the following locations:
+ *
+ * http://www.opensource.org/licenses/gpl-license.html
+ * http://www.gnu.org/copyleft/gpl.html
+ */
+
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	chosen {
+		stdout-path = &uart2;
+	};
+
+	memory {
+		reg = <0x10000000 0x40000000>;
+	};
+
+	aliases {
+		can0 = &can1;
+		can1 = &can2;
+		led0 = &led0;
+		led1 = &led1;
+		led2 = &led2;
+		nand = &gpmi;
+		usb0 = &usbh1;
+		usb1 = &usbotg;
+		rtc0 = &i2c_rtc;
+		rtc1 = &snvs;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led0: D1 {
+			label = "D1";
+			gpios = <&gpio1 2 0>;
+			default-state = "on";
+			linux,default-trigger = "heartbeat";
+		};
+
+		led1: D2 {
+			label = "D2";
+			gpios = <&gpio1 0 0>;
+			default-state = "off";
+		};
+
+		led2: D3 {
+                        label = "D3";
+                        gpios = <&gpio1 4 0>;
+                        default-state = "on";
+                };
+	};
+
+	skov_version {
+		pinctrl-names = "default";
+		pinctrl-0 = <&pinctrl_skov_version>;
+		compatible = "skov-version";
+		v0-gpio = <&gpio2 10 0>;
+		v1-gpio = <&gpio2 15 0>;
+		v2-gpio = <&gpio2 8 0>;
+		v3-gpio = <&gpio2 9 0>;
+		v4-gpio = <&gpio2 12 0>;
+		v5-gpio = <&gpio2 14 0>;
+		v6-gpio = <&gpio2 11 0>;
+		v7-gpio = <&gpio2 13 0>;
+	};
+
+	regulators {
+		compatible = "simple-bus";
+
+		/* power in from 24 V */
+		reg_24p0v: 24p0v {
+			compatible = "regulator-fixed";
+			regulator-name = "24P0V";
+			regulator-min-microvolt = <24000000>;
+			regulator-max-microvolt = <24000000>;
+			regulator-always-on;
+		};
+
+		reg_5p0v: 5p0v {
+			compatible = "regulator-fixed";
+			regulator-name = "5P0V";
+			regulator-min-microvolt = <5000000>;
+			regulator-max-microvolt = <5000000>;
+			regulator-always-on;
+		};
+
+		reg_1p2v: 1p2v {
+			compatible = "regulator-fixed";
+			vin-supply = <&reg_5p0v>;
+			regulator-name = "1P2V";
+			regulator-min-microvolt = <1200000>;
+			regulator-max-microvolt = <1200000>;
+			regulator-always-on;
+		};
+
+		reg_1p5v: 1p5v {
+			compatible = "regulator-fixed";
+			vin-supply = <&reg_5p0v>;
+			regulator-name = "1P5V";
+			regulator-min-microvolt = <1500000>;
+			regulator-max-microvolt = <1500000>;
+			regulator-always-on;
+		};
+
+		reg_2p5v: 2p5v {
+			compatible = "regulator-fixed";
+			vin-supply = <&reg_5p0v>;
+			regulator-name = "2P5V";
+			regulator-min-microvolt = <2500000>;
+			regulator-max-microvolt = <2500000>;
+			regulator-always-on;
+		};
+
+		reg_3p3v: 3p3v {
+			compatible = "regulator-fixed";
+			vin-supply = <&reg_5p0v>;
+			regulator-name = "3P3V";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			regulator-always-on;
+		};
+
+		reg_vref: vref {
+			compatible = "regulator-fixed";
+			regulator-name = "vref";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			regulator-always-on;
+		};
+
+		reg_usb_otg_vbus: usb_otg_vbus {
+			compatible = "regulator-fixed";
+			vin-supply = <&reg_5p0v>;
+			regulator-name = "usb_otg_vbus";
+			regulator-min-microvolt = <5000000>;
+			regulator-max-microvolt = <5000000>;
+			regulator-always-on;
+		};
+
+		reg_usb_h1_vbus: usb_h1_vbus {
+			compatible = "regulator-fixed";
+			vin-supply = <&reg_5p0v>;
+			regulator-name = "usb_h1_vbus";
+			regulator-min-microvolt = <5000000>;
+			regulator-max-microvolt = <5000000>;
+			regulator-always-on;
+		};
+
+		/* MMC power supply */
+		vcc_mmc: sdcard_power_supply {
+			compatible = "regulator-fixed";
+			vin-supply = <&reg_3p3v>;
+			regulator-name = "mmc_vcc_supply";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+		};
+
+		/* MMC IO voltage select */
+		iovcc_mmc: sdcard_io_supply {
+			pinctrl-names = "default";
+			pinctrl-0 = <&pinctrl_io_vcc_mmc>;
+			compatible = "regulator-gpio";
+			vin-supply = <&reg_5p0v>;
+			regulator-name = "mmc_io_supply";
+			regulator-type = "voltage";
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <3300000>;
+			gpios = <&gpio7 13 GPIO_ACTIVE_HIGH>;
+			states = <1800000 0x1
+				  3300000 0x0>;
+			startup-delay-us = <100>;
+		};
+	};
+
+	display {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		status = "okay";
+		compatible = "fsl,imx-parallel-display";
+		pinctrl-names = "default";
+		pinctrl-0 = <&pinctrl_ipu1_4>;
+
+		port@0 {
+			reg = <0>;
+
+			display0_in: endpoint {
+				remote-endpoint = <&ipu1_di0_disp0>;
+			};
+		};
+
+		port@1 {
+			reg = <1>;
+
+			display0_out: endpoint {
+				remote-endpoint = <&panel_in>;
+			};
+		};
+	};
+
+	backlight_lcd: backlight_lcd {
+		compatible = "pwm-backlight";
+		pwms = <&pwm2 0 20000>;
+		brightness-levels = <0 16 32 48 64 80 96 112 128 144 160 176 192 208 224 240 255>;
+		default-brightness-level = <8>;
+		power-supply = <&reg_24p0v>;
+		status = "okay";
+	};
+
+	panel: panel {
+		compatible = "logictechnologies,lttd800480070-l2rt";
+		backlight = <&backlight_lcd>;
+		power-supply = <&reg_3p3v>;
+
+		port {
+			panel_in: endpoint {
+				remote-endpoint = <&display0_out>;
+			};
+		};
+	};
+};
+
+&ipu1_di0_disp0 {
+	remote-endpoint = <&display0_in>;
+};
+
+&ecspi1 {
+	fsl,spi-num-chipselects = <1>;
+	cs-gpios = <&gpio3 24 0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_ecspi1_1>;
+	status = "okay";
+
+	flash: m25p80@0 {
+		compatible = "n25q064";
+		spi-max-frequency = <54000000>;
+		reg = <0>;
+		#address-cells = <1>;
+		#size-cells = <1>;
+	};
+};
+
+&ecspi2 {
+	fsl,spi-num-chipselects = <1>;
+	cs-gpios = <&gpio2 26 0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_ecspi2_1>;
+	status = "okay";
+
+	mcp3x0x@0 {
+		compatible = "mcp3002";
+		reg = <0>;
+		spi-max-frequency = <1000000>;
+	};
+};
+
+&ecspi4 {
+	fsl,spi-num-chipselects = <1>;
+	cs-gpios = <&gpio3 20 0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_ecspi4_1>;
+	status = "okay";
+
+	ads7846: ads7846@0 {
+		compatible = "ti,tsc2046";
+		reg = <0>;
+		spi-max-frequency = <1000000>;
+		vcc-supply  = <&reg_3p3v>;
+
+		ti,x-min = /bits/ 16 <800>;
+		ti,x-max = /bits/ 16 <3400>;
+		ti,y-min = /bits/ 16 <500>;
+		ti,y-max = /bits/ 16 <3400>;
+		ti,x-plate-ohms = /bits/ 16 <642>;
+		ti,y-plate-ohms = /bits/ 16 <295>;
+		ti,pressure-max = /bits/ 16 <1500>;
+		ti,vref-delay-usecs = /bits/ 16 <300>;
+		ti,debounce-max = /bits/ 16 <100>;
+		ti,debounce-tol = /bits/ 16 <(~0)>;
+		ti,debounce-rep = /bits/ 16 <8>;
+
+		linux,wakeup;
+	};
+};
+
+&fec {
+	compatible = "fsl,imx6q-fec";
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_enet_4>;
+	phy-mode = "rmii";
+	phy-reset-gpios = <&gpio1 5 0>;
+	phy-reset-duration = "100";
+	phy-supply = <&reg_3p3v>;
+	#address-cells = <0>;
+	#size-cells = <1>;
+
+	/*
+	 * MDIO phy#1 is port 1 e.g. LAN1
+	 * MDIO phy#2 is port 2 e.g. LAN2
+	 * MDIO phy#2 is port 2 e.g. RMII port, with no real and usable
+	 * phy emulation
+	 */
+	fixed-link {
+		speed = <100>;
+		full-duplex;
+	};
+};
+
+/* signal "active link" on the RMII to the ethernet switch */
+&gpio1 {
+	active_rmii_link {
+		gpio-hog;
+		gpios = <&gpio1 24 GPIO_ACTIVE_LOW>;
+		output-low;
+		line-name = "rmii-active-link";
+	};
+};
+
+&i2c2 {
+	clock-frequency = <100000>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_i2c2_2>;
+};
+
+&i2c3 {
+	status = "okay";
+	clock-frequency = <400000>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_i2c3_2>;
+
+	i2c_rtc: pcf85063@51 {
+		compatible = "nxp,pcf85063";
+		reg = <0x51>;
+		quartz-load-femtofarad = <12500>;
+	};
+};
+
+&gpmi {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_gpmi_nand_2>;
+	nand-on-flash-bbt;
+	status = "okay";
+};
+
+&iomuxc {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_hog>;
+
+	pinctrl_board: skov_imx6_grp {
+		pinctrl_hog: hoggrp {
+			fsl,pins = <
+				/* external pull up */
+				MX6QDL_PAD_EIM_D30__GPIO3_IO30		0xb0		/* ethernet phy interrupt */
+			>;
+		};
+
+		pinctrl_skov_version: skov_version_pins {
+			fsl,pins = <
+				/* never changing input pins */
+				MX6QDL_PAD_SD4_DAT0__GPIO2_IO08		0x1b040		/* VC0 */
+				MX6QDL_PAD_SD4_DAT1__GPIO2_IO09		0x1b040		/* VC1 */
+				MX6QDL_PAD_SD4_DAT2__GPIO2_IO10		0x1b040		/* VC2 */
+				MX6QDL_PAD_SD4_DAT3__GPIO2_IO11		0x1b040		/* VC3 */
+				MX6QDL_PAD_SD4_DAT4__GPIO2_IO12		0x1b040		/* VC4 */
+				MX6QDL_PAD_SD4_DAT5__GPIO2_IO13		0x1b040		/* VC5 */
+				MX6QDL_PAD_SD4_DAT6__GPIO2_IO14		0x1b040		/* VC6 */
+				MX6QDL_PAD_SD4_DAT7__GPIO2_IO15		0x1b040		/* VC7 */
+			>;
+		};
+
+		pinctrl_io_vcc_mmc: vcc_io_mmc_grp {
+			fsl,pins = <
+				MX6QDL_PAD_GPIO_18__GPIO7_IO13		0x40000058	/* MMC IO Supply Select */
+			>;
+		};
+
+		pinctrl_ecspi1_1: ecspi1grp-1 {
+			fsl,pins = <
+				MX6QDL_PAD_EIM_D17__ECSPI1_MISO		0x100b1
+				MX6QDL_PAD_EIM_D18__ECSPI1_MOSI		0x000b1
+				MX6QDL_PAD_EIM_D16__ECSPI1_SCLK		0x000b1
+				/* *no* external pull up */
+				MX6QDL_PAD_EIM_D24__GPIO3_IO24		0x40000058	/* spi-nor CS */
+			>;
+		};
+
+		pinctrl_ecspi2_1: ecspi2grp-1 {
+			fsl,pins = <
+				MX6QDL_PAD_EIM_OE__ECSPI2_MISO		0x100b1
+				MX6QDL_PAD_EIM_CS1__ECSPI2_MOSI		0x000b1
+				MX6QDL_PAD_EIM_CS0__ECSPI2_SCLK 	0x000b1
+				/* external pull up */
+				MX6QDL_PAD_EIM_RW__GPIO2_IO26		0x40000058	/* mcp2003 CS */
+			>;
+		};
+
+		pinctrl_ecspi4_1: ecspi4grp-1 {
+			fsl,pins = <
+				MX6QDL_PAD_EIM_D28__ECSPI4_MOSI 	0x100b1
+				MX6QDL_PAD_EIM_D22__ECSPI4_MISO 	0x000b1
+				MX6QDL_PAD_EIM_D21__ECSPI4_SCLK 	0x000b1
+				/* *no* external pull up */
+				MX6QDL_PAD_EIM_D20__GPIO3_IO20		0x40000058	/* TSC2046 CS */
+			>;
+		};
+
+		pinctrl_enet_4: enetgrp-4 {
+			fsl,pins = <
+				/* MDIO speed up to 5 MHz */
+				MX6QDL_PAD_ENET_MDIO__ENET_MDIO		0x100b1
+				MX6QDL_PAD_ENET_MDC__ENET_MDC		0x000b1
+				/* RMII 50 MHz */
+				MX6QDL_PAD_ENET_CRS_DV__ENET_RX_EN	0x100f5
+				MX6QDL_PAD_ENET_TX_EN__ENET_TX_EN	0x100f5
+				MX6QDL_PAD_ENET_RXD0__ENET_RX_DATA0	0x100c0
+				MX6QDL_PAD_ENET_RXD1__ENET_RX_DATA1	0x100c0
+				MX6QDL_PAD_ENET_TXD0__ENET_TX_DATA0	0x100f5
+				MX6QDL_PAD_ENET_TXD1__ENET_TX_DATA1	0x100f5
+				MX6QDL_PAD_GPIO_16__ENET_REF_CLK	0x40010040
+				MX6QDL_PAD_GPIO_5__GPIO1_IO05		0x40000058
+				/* GPIO for "link active" */
+				MX6QDL_PAD_ENET_RX_ER__GPIO1_IO24	0x40000058
+			>;
+		};
+
+		pinctrl_pwm2_2: pwm2grp-2 {
+			fsl,pins = <
+				MX6QDL_PAD_GPIO_1__PWM2_OUT		0x00058
+			>;
+		};
+
+		pinctrl_pwm3_2: pwm3grp-2 {
+			fsl,pins = <
+				MX6QDL_PAD_SD1_DAT1__PWM3_OUT		0x00058
+			>;
+		};
+
+		pinctrl_flexcan1_3: flexcan1grp-3 {
+			fsl,pins = <
+				MX6QDL_PAD_GPIO_7__FLEXCAN1_TX		0x80000000
+				MX6QDL_PAD_GPIO_8__FLEXCAN1_RX		0x80000000
+				MX6QDL_PAD_EIM_D31__GPIO3_IO31		0x13000		/* CAN RS */
+			>;
+		};
+
+		pinctrl_usdhc3_2: usdhc3grp-2 {
+			fsl,pins = <
+				/* SoC internal pull up required */
+				MX6QDL_PAD_SD3_CMD__SD3_CMD		0x17059
+				MX6QDL_PAD_SD3_CLK__SD3_CLK		0x10059
+				MX6QDL_PAD_SD3_DAT0__SD3_DATA0		0x17059
+				MX6QDL_PAD_SD3_DAT1__SD3_DATA1		0x17059
+				MX6QDL_PAD_SD3_DAT2__SD3_DATA2		0x17059
+				MX6QDL_PAD_SD3_DAT3__SD3_DATA3		0x17059
+				/* SoC internal pull up required */
+				MX6QDL_PAD_SD3_DAT4__GPIO7_IO01		0x4001b040	/* WP */
+				/* SoC internal pull up required */
+				MX6QDL_PAD_SD3_DAT5__GPIO7_IO00		0x4001b040	/* CD */
+			>;
+		};
+
+		pinctrl_usdhc3_100mhz: usdhc3grp100mhz {
+			fsl,pins = <
+				MX6QDL_PAD_SD3_CMD__SD3_CMD		0x170b9
+				MX6QDL_PAD_SD3_CLK__SD3_CLK		0x100b9
+				MX6QDL_PAD_SD3_DAT0__SD3_DATA0		0x170b9
+				MX6QDL_PAD_SD3_DAT1__SD3_DATA1		0x170b9
+				MX6QDL_PAD_SD3_DAT2__SD3_DATA2		0x170b9
+				MX6QDL_PAD_SD3_DAT3__SD3_DATA3		0x170b9
+				MX6QDL_PAD_SD3_DAT4__GPIO7_IO01		0x4001b040
+				MX6QDL_PAD_SD3_DAT5__GPIO7_IO00		0x4001b040
+			>;
+		};
+
+		pinctrl_usdhc3_200mhz: usdhc3grp200mhz {
+			fsl,pins = <
+				MX6QDL_PAD_SD3_CMD__SD3_CMD		0x170f9
+				MX6QDL_PAD_SD3_CLK__SD3_CLK		0x100f9
+				MX6QDL_PAD_SD3_DAT0__SD3_DATA0		0x170f9
+				MX6QDL_PAD_SD3_DAT1__SD3_DATA1		0x170f9
+				MX6QDL_PAD_SD3_DAT2__SD3_DATA2		0x170f9
+				MX6QDL_PAD_SD3_DAT3__SD3_DATA3		0x170f9
+				MX6QDL_PAD_SD3_DAT4__GPIO7_IO01		0x4001b040
+				MX6QDL_PAD_SD3_DAT5__GPIO7_IO00		0x4001b040
+			>;
+		};
+
+		pinctrl_i2c2_2: i2c2grp-2 {
+			fsl,pins = <
+				/* internal 22 k pull up required */
+				MX6QDL_PAD_KEY_COL3__I2C2_SCL		0x4001F878
+				/* internal 22 k pull up required */
+				MX6QDL_PAD_KEY_ROW3__I2C2_SDA		0x4001F878
+			>;
+		};
+
+		pinctrl_i2c3_2: i2c3grp-2 {
+			fsl,pins = <
+				/* external 10 k pull up */
+				MX6QDL_PAD_GPIO_3__I2C3_SCL		0x40010878
+				/* external 10 k pull up */
+				MX6QDL_PAD_GPIO_6__I2C3_SDA		0x40010878
+			>;
+		};
+
+		pinctrl_ipu1_4: ipu1grp-4 {
+			fsl,pins = <
+				MX6QDL_PAD_DI0_DISP_CLK__IPU1_DI0_DISP_CLK	0x10
+				MX6QDL_PAD_DI0_PIN15__IPU1_DI0_PIN15		0x10
+				MX6QDL_PAD_DI0_PIN2__IPU1_DI0_PIN02		0x10
+				MX6QDL_PAD_DI0_PIN3__IPU1_DI0_PIN03		0x10
+				MX6QDL_PAD_DISP0_DAT0__IPU1_DISP0_DATA00	0x10
+				MX6QDL_PAD_DISP0_DAT1__IPU1_DISP0_DATA01	0x10
+				MX6QDL_PAD_DISP0_DAT2__IPU1_DISP0_DATA02	0x10
+				MX6QDL_PAD_DISP0_DAT3__IPU1_DISP0_DATA03	0x10
+				MX6QDL_PAD_DISP0_DAT4__IPU1_DISP0_DATA04	0x10
+				MX6QDL_PAD_DISP0_DAT5__IPU1_DISP0_DATA05	0x10
+				MX6QDL_PAD_DISP0_DAT6__IPU1_DISP0_DATA06	0x10
+				MX6QDL_PAD_DISP0_DAT7__IPU1_DISP0_DATA07	0x10
+				MX6QDL_PAD_DISP0_DAT8__IPU1_DISP0_DATA08	0x10
+				MX6QDL_PAD_DISP0_DAT9__IPU1_DISP0_DATA09	0x10
+				MX6QDL_PAD_DISP0_DAT10__IPU1_DISP0_DATA10	0x10
+				MX6QDL_PAD_DISP0_DAT11__IPU1_DISP0_DATA11	0x10
+				MX6QDL_PAD_DISP0_DAT12__IPU1_DISP0_DATA12	0x10
+				MX6QDL_PAD_DISP0_DAT13__IPU1_DISP0_DATA13	0x10
+				MX6QDL_PAD_DISP0_DAT14__IPU1_DISP0_DATA14	0x10
+				MX6QDL_PAD_DISP0_DAT15__IPU1_DISP0_DATA15	0x10
+				MX6QDL_PAD_DISP0_DAT16__IPU1_DISP0_DATA16	0x10
+				MX6QDL_PAD_DISP0_DAT17__IPU1_DISP0_DATA17	0x10
+				MX6QDL_PAD_DISP0_DAT18__IPU1_DISP0_DATA18	0x10
+				MX6QDL_PAD_DISP0_DAT19__IPU1_DISP0_DATA19	0x10
+				MX6QDL_PAD_DISP0_DAT20__IPU1_DISP0_DATA20	0x10
+				MX6QDL_PAD_DISP0_DAT21__IPU1_DISP0_DATA21	0x10
+				MX6QDL_PAD_DISP0_DAT22__IPU1_DISP0_DATA22	0x10
+				MX6QDL_PAD_DISP0_DAT23__IPU1_DISP0_DATA23	0x10
+			>;
+		};
+
+		pinctrl_uart2_1: uart2grp-1 {
+			fsl,pins = <
+				MX6QDL_PAD_EIM_D26__UART2_TX_DATA	0x1b0b1
+				MX6QDL_PAD_EIM_D27__UART2_RX_DATA	0x1b0b1
+			>;
+		};
+
+		pinctrl_gpmi_nand_2: gpmi-nand-2 {
+			fsl,pins = <
+				MX6QDL_PAD_NANDF_CLE__NAND_CLE		0xb0b1
+				MX6QDL_PAD_NANDF_ALE__NAND_ALE		0xb0b1
+				MX6QDL_PAD_NANDF_RB0__NAND_READY_B	0xb000
+				MX6QDL_PAD_NANDF_CS0__NAND_CE0_B	0xb0b1
+				MX6QDL_PAD_NANDF_CS1__NAND_CE1_B	0xb0b1
+				MX6QDL_PAD_SD4_CMD__NAND_RE_B		0xb0b1
+				MX6QDL_PAD_SD4_CLK__NAND_WE_B		0xb0b1
+				MX6QDL_PAD_NANDF_D0__NAND_DATA00	0xb0b1
+				MX6QDL_PAD_NANDF_D1__NAND_DATA01	0xb0b1
+				MX6QDL_PAD_NANDF_D2__NAND_DATA02	0xb0b1
+				MX6QDL_PAD_NANDF_D3__NAND_DATA03	0xb0b1
+				MX6QDL_PAD_NANDF_D4__NAND_DATA04	0xb0b1
+				MX6QDL_PAD_NANDF_D5__NAND_DATA05	0xb0b1
+				MX6QDL_PAD_NANDF_D6__NAND_DATA06	0xb0b1
+				MX6QDL_PAD_NANDF_D7__NAND_DATA07	0xb0b1
+			>;
+		};
+	};
+};
+
+&pwm2 {
+	/* used for backlight brightness */
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_pwm2_2>;
+	status = "okay";
+};
+
+&pwm3 {
+	/* used for LCD contrast control */
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_pwm3_2>;
+	status = "okay";
+};
+
+&can1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_flexcan1_3>;
+	xceiver-supply = <&reg_3p3v>;
+	status = "okay";
+};
+
+&usbh1 {
+	vbus-supply = <&reg_usb_h1_vbus>;
+	disable-over-current;
+        status = "okay";
+};
+
+&usbotg {
+	vbus-supply = <&reg_usb_otg_vbus>;
+	disable-over-current;
+	status = "okay";
+};
+
+&uart2 {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart2_1>;
+};
+
+&usdhc3 {
+	pinctrl-names = "default", "state_100mhz", "state_200mhz";
+	pinctrl-0 = <&pinctrl_usdhc3_2>;
+	pinctrl-1 = <&pinctrl_usdhc3_100mhz>;
+	pinctrl-2 = <&pinctrl_usdhc3_200mhz>;
+	wp-gpios = <&gpio7 1 GPIO_ACTIVE_HIGH>; /* e.g. low if not 'locked' */
+	cd-gpios = <&gpio7 0 GPIO_ACTIVE_LOW>; /* e.g. high if no card is plugged in */
+	status = "okay";
+	cap-power-off-card;
+	full-pwr-cycle;
+	bus-width = <4>;
+	max-frequency = <50000000>;
+	cap-sd-highspeed;
+	sd-uhs-sdr12;
+	sd-uhs-sdr25;
+	sd-uhs-sdr50;
+	sd-uhs-sdr104;
+	sd-uhs-ddr50;
+	mmc-ddr-1_8v;
+	mmc-hs200-1_8v;
+	mmc-hs400-1_8v;
+	vmmc-supply = <&vcc_mmc>;
+	vqmmc-supply = <&iovcc_mmc>;
+	fsl,delay-line;
+};
+
+&ocotp {
+	read-only;
+};
